src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs

```diff
@@ -128,6 +147,48 @@ public ExecutionStatus Approve(UInt160 spender, UInt256 amount, SystemContractEx
             frame.ReturnValue = 1.ToUInt256().ToBytes();
             return ExecutionStatus.Ok;
         }
+        
+        [ContractMethod(Lrc20Interface.MethodSetAllowedSupply)]
+        public ExecutionStatus SetAllowedSupply(Money amount, SystemContractExecutionFrame frame)
+        {
+            frame.UseGas(GasMetering.NativeTokenApproveCost);
+            if(!frame.InvocationContext.Sender.Equals(_mintCntrlAdd))
+                return ExecutionStatus.ExecutionHalted;
+            
+            if(_allowedSupply + amount > _maxSupply)
+                return ExecutionStatus.ExecutionHalted;
+            
+            _allowedSupply = amount;
+            frame.ReturnValue = _allowedSupply.ToUInt256().ToBytes();
+            return ExecutionStatus.Ok;
+        }
+        
+        [ContractMethod(Lrc20Interface.MethodGetAllowedSupply)]
+        public ExecutionStatus GetAllowedSupply(SystemContractExecutionFrame frame)
+        {
+            frame.UseGas(GasMetering.NativeTokenApproveCost);
+            frame.ReturnValue = _allowedSupply.ToUInt256().ToBytes();
+            return ExecutionStatus.Ok;
+        }
+        
+        [ContractMethod(Lrc20Interface.MethodMint)]
+        public ExecutionStatus Mint(UInt160 address, Money amount, SystemContractExecutionFrame frame)
+        {
+            frame.UseGas(GasMetering.NativeTokenApproveCost);
+            if(!frame.InvocationContext.Sender.Equals(_minterAdd))
+                return ExecutionStatus.ExecutionHalted;
+            
+            var currentBalance =  _context.Snapshot?.Balances.GetBalance(address);
+            if (currentBalance is null) return ExecutionStatus.ExecutionHalted;
+            
+            if(currentBalance + amount > _maxSupply || currentBalance + amount > _allowedSupply)
```

Comment:
We should check totalSupply here,  not current address balance

Reply:
incorporated